From 2eef4e3d2aa1838849343757d1b2e24d28f0d554 Mon Sep 17 00:00:00 2001
From: Sam Nazarko <email@samnazarko.co.uk>
Date: Tue, 7 Jul 2020 06:09:35 +0200
Subject: [PATCH 16/41] AMLUtils: add new resolution parsing methods to support

---
 xbmc/utils/AMLUtils.cpp | 59 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 59 insertions(+)

diff --git a/xbmc/utils/AMLUtils.cpp b/xbmc/utils/AMLUtils.cpp
index a4aa0924d3..95c61c78d7 100644
--- a/xbmc/utils/AMLUtils.cpp
+++ b/xbmc/utils/AMLUtils.cpp
@@ -551,6 +551,65 @@ bool aml_set_native_resolution(const RESOLUTION_INFO &res, std::string framebuff
   return result;
 }
 
+static void aml_add_resolution(const std::string &mode, std::vector<RESOLUTION_INFO> &resolutions)
+{
+  RESOLUTION_INFO res;
+
+  if (((StringUtils::StartsWith(mode, "4k2k")) && (aml_support_h264_4k2k() > AML_NO_H264_4K2K)) || !(StringUtils::StartsWith(mode, "4k2k")))
+  {
+    if (aml_mode_to_resolution(mode.c_str(), &res))
+      resolutions.push_back(res);
+
+    if (aml_has_frac_rate_policy())
+    {
+      // Add fractional frame rates: 23.976, 29.97 and 59.94 Hz
+      switch ((int)res.fRefreshRate)
+      {
+        case 24:
+        case 30:
+        case 60:
+        res.fRefreshRate /= 1.001;
+        res.strMode       = StringUtils::Format("%dx%d @ %.2f%s - Full Screen", res.iScreenWidth, res.iScreenHeight, res.fRefreshRate,
+            res.dwFlags & D3DPRESENTFLAG_INTERLACED ? "i" : "");
+        resolutions.push_back(res);
+      }
+    }
+  }
+}
+
+static bool aml_modes_to_resolutions(const std::vector<std::string> &probe_str, std::vector<RESOLUTION_INFO> &resolutions)
+{
+  resolutions.clear();
+
+  for (std::vector<std::string>::const_iterator i = probe_str.begin(); i != probe_str.end(); ++i)
+  {
+    std::vector<std::string> elem_str = StringUtils::Split(*i, " ");
+    if (elem_str.size() == 1)
+      aml_add_resolution(elem_str[0], resolutions);
+    else
+      for (size_t j = 1; j < elem_str.size(); j++)
+        aml_add_resolution(elem_str[0] + " " + elem_str[j], resolutions);
+  }
+  
+  return resolutions.size() > 0;
+}
+
+bool aml_probe_3d_resolutions(std::vector<RESOLUTION_INFO> &resolutions)
+{
+  std::string valstr, dcap3dfile;
+
+  dcap3dfile = CSpecialProtocol::TranslatePath("special://home/userdata/disp_cap_3d");
+
+  if (SysfsUtils::GetString(dcap3dfile, valstr) >= 0 ||
+      SysfsUtils::GetString("/sys/class/amhdmitx/amhdmitx0/disp_cap_3d", valstr) >= 0)
+  {
+    std::vector<std::string> probe_str = StringUtils::Split(valstr, "\n");
+    return aml_modes_to_resolutions(probe_str, resolutions);
+  }
+
+  return false;
+}
+
 bool aml_probe_resolutions(std::vector<RESOLUTION_INFO> &resolutions)
 {
   std::string valstr, addstr, dcapfile, daddfile;
-- 
2.28.0

