From 8761278bcef61abaf198a570a308bce712744f04 Mon Sep 17 00:00:00 2001
From: Sam Nazarko <email@samnazarko.co.uk>
Date: Sat, 3 Aug 2019 21:10:00 +0100
Subject: [PATCH 17/41] AMLUtils: use new 3D parsing method

Signed-off-by: Sam Nazarko <email@samnazarko.co.uk>
---
 xbmc/utils/AMLUtils.cpp | 28 +---------------------------
 1 file changed, 1 insertion(+), 27 deletions(-)

diff --git a/xbmc/utils/AMLUtils.cpp b/xbmc/utils/AMLUtils.cpp
index 95c61c78d7..63e1624c3b 100644
--- a/xbmc/utils/AMLUtils.cpp
+++ b/xbmc/utils/AMLUtils.cpp
@@ -638,33 +638,7 @@ bool aml_probe_resolutions(std::vector<RESOLUTION_INFO> &resolutions)
   }
   std::vector<std::string> probe_str = StringUtils::Split(valstr, "\n");
 
-  resolutions.clear();
-  RESOLUTION_INFO res;
-  for (std::vector<std::string>::const_iterator i = probe_str.begin(); i != probe_str.end(); ++i)
-  {
-    if (((StringUtils::StartsWith(i->c_str(), "4k2k")) && (aml_support_h264_4k2k() > AML_NO_H264_4K2K)) || !(StringUtils::StartsWith(i->c_str(), "4k2k")))
-    {
-      if (aml_mode_to_resolution(i->c_str(), &res))
-        resolutions.push_back(res);
-
-      if (aml_has_frac_rate_policy())
-      {
-        // Add fractional frame rates: 23.976, 29.97 and 59.94 Hz
-        switch ((int)res.fRefreshRate)
-        {
-          case 24:
-          case 30:
-          case 60:
-            res.fRefreshRate /= 1.001;
-            res.strMode       = StringUtils::Format("%dx%d @ %.2f%s - Full Screen", res.iScreenWidth, res.iScreenHeight, res.fRefreshRate,
-              res.dwFlags & D3DPRESENTFLAG_INTERLACED ? "i" : "");
-            resolutions.push_back(res);
-            break;
-        }
-      }
-    }
-  }
-  return resolutions.size() > 0;
+  return aml_modes_to_resolutions(probe_str, resolutions);
 }
 
 bool aml_get_preferred_resolution(RESOLUTION_INFO *res)
-- 
2.28.0

